### ISSUE TITLE
correct serious mistakes in tests/Makefile, patch included

### ISSUE THREAD

**Issue by binnacle** [facebook/zstd#985]
"legacy" test was failing, several other targets also have problem

[zstd-20180113_patch.txt](https://github.com/facebook/zstd/files/1628638/zstd-20180113_patch.txt)
[zstd-v1.3.3_patch.txt](https://github.com/facebook/zstd/files/1628639/zstd-v1.3.3_patch.txt)

**Comment by Cyan4973** [comment]
I'm not sure to understand this issue.
Your patches mostly repeat flags already present within `CFLAGS` or `CPPFLAGS`, and add them to `LDFLAGS`. But since the linking command is just below, with all 3 `CFLAGS` `CPPFLAGS` and `LDFLAGS` already included (within `FLAGS`), the main impact is to double mention of these flags on compilation line, which produces no effect.
So it's not clear to me what these patches were solving for you.

I've also tests several targets in `tests/Makefile`, and they all seem to work on my test systems (one Mac OS-X, one Linux Ubuntu). Could you please provide more details on your configuration, and the tests which fail, and how they fail ?

**Comment by binnacle** [comment]
Appears to be a difference between GNU make and whatever make you use.  GNU make appears to complete all variable expansions in its "first phase," so that "target specific rules" do not benefit from dynamic expansion.  Running version 4.1 and didn't notice anything in the 4.2 release notes indicating a feature change in this area.

https://stackoverflow.com/questions/4069515/bug-in-gnu-make-target-specific-variables-are-not-expanded-in-implicit-rules

patch above is not correct then for portable builds, requires something like

https://www.gnu.org/software/make/manual/make.html#Secondary-Expansion

or IMO more cleanly by writing separate object and executable targets for the test programs

**Comment by binnacle** [comment]
If the intent of commit f8a05932b9bfd504614b15b8898eaaa3e3b8520b was to correct this issue, it was ineffective:

.
.
.
gcc -I../lib -I../lib/common -I../lib/compress -I../lib/dictBuilder -I../lib/deprecated -I../programs -I../lib/legacy -O3 -flto -g -I/usr/local/zlib64/include -I/usr/local/include -L/usr/local/zlib64/lib -L/usr/local/lib64 -L/usr/local/lib  ../lib/decompress/zstd_decompress.c ../lib/decompress/huf_decompress.c ../lib/common/fse_decompress.c ../lib/common/entropy_common.c ../lib/common/pool.c ../lib/common/error_private.c ../lib/common/xxhash.c ../lib/common/threading.c ../lib/common/zstd_common.c ../lib/compress/zstd_double_fast.c ../lib/compress/zstd_lazy.c ../lib/compress/zstd_fast.c ../lib/compress/zstd_ldm.c ../lib/compress/huf_compress.c ../lib/compress/fse_compress.c ../lib/compress/zstd_compress.c ../lib/compress/zstdmt_compress.c ../lib/compress/zstd_opt.c ../lib/legacy/zstd_v07.c ../lib/legacy/zstd_v06.c ../lib/legacy/zstd_v05.c ../lib/legacy/zstd_v03.c ../lib/legacy/zstd_v01.c ../lib/legacy/zstd_v04.c ../lib/legacy/zstd_v02.c legacy.c -o legacy
./legacy
ERROR: Unknown frame descriptor
Makefile:386: recipe for target 'test-legacy' failed
make[1]: *** [test-legacy] Error 1
make[1]: Leaving directory '/w/gpl/zstd-dev_20180114/tests'
Makefile:80: recipe for target 'test' failed
make: *** [test] Error 2

**Comment by Cyan4973** [comment]
nope.
The intent of this patch was only to add `list` target to the `Makefile`.

The compilation line listed in your post is missing the `-DZSTD_LEGACY_SUPPORT=4` macro definition, which is important to enable legacy support. Without it, legacy support is simply not present, resulting in a test error.
The compilation line has a few other elements missing (many warning flags) or added (several include paths), so it seems it was modified, and/or interpreted differently.

Here is what is displayed on my system : 
```bash
cc -I../lib -I../lib/common -I../lib/compress -I../lib/dictBuilder -I../lib/deprecated -I../programs -I../lib/legacy -O3 -Wall -Wextra -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wformat-security -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -g -DZSTD_DEBUG=1  -DZSTD_LEGACY_SUPPORT=4  ../lib/decompress/huf_decompress.c ../lib/decompress/zstd_decompress.c ../lib/common/entropy_common.c ../lib/common/error_private.c ../lib/common/fse_decompress.c ../lib/common/pool.c ../lib/common/threading.c ../lib/common/xxhash.c ../lib/common/zstd_common.c ../lib/compress/fse_compress.c ../lib/compress/huf_compress.c ../lib/compress/zstd_compress.c ../lib/compress/zstd_double_fast.c ../lib/compress/zstd_fast.c ../lib/compress/zstd_lazy.c ../lib/compress/zstd_ldm.c ../lib/compress/zstd_opt.c ../lib/compress/zstdmt_compress.c ../lib/legacy/zstd_v01.c ../lib/legacy/zstd_v02.c ../lib/legacy/zstd_v03.c ../lib/legacy/zstd_v04.c ../lib/legacy/zstd_v05.c ../lib/legacy/zstd_v06.c ../lib/legacy/zstd_v07.c legacy.c -o legacy
```

Do you invoke `make` with your own `CFLAGS` argument ?

**Comment by binnacle** [comment]
yes

CFLAGS="-O3 -g -I/usr/local/include -L/usr/local/lib64 -L/usr/local/lib"

**Comment by Cyan4973** [comment]
`-O3 -g` is fine for `CFLAGS`.
`-I/usr/local/include` rather belongs to `CPPFLAGS`.
`-L/usr/local/lib64 -L/usr/local/lib` better belong to `LDFLAGS`.

More importantly, you don't specify if you use this assignment before or after `make` in the command line.
It makes a large difference.

Use them before `make` if you want these values to be included as part of the `Makefile` script.
Use them after `make` if you want to __overwrite__ any related assignment in `Makefile`.
This second solution is not recommended, it means you take full ownership of these variables, which requires to know exactly what you are preventing to be set. 
I guess this is exactly what's happening in your scenario.

In conclusion, try to run the following command : 
```
CFLAGS="-O3 -g" CPPFLAGS=-I/usr/local/include LDFLAGS="-L/usr/local/lib64 -L/usr/local/lib" make test-legacy
```

**Comment by binnacle** [comment]
That's a subtle bit of knowledge I've managed to get by without for many years
--thank you for identifying the issue and the explanation.

I have observations:

Anyone conditioned  to building with GNU auto* tools will biased toward the command-parameter approach rather than the inherited environment approach due to the --help text 'configure' displays.  The "overwrite" approach does not result in side-effects in this case.

While clever, due to the aforementioned the code in tests/Makefile is fragile relative to common usage.

This deserves mention in the README file.
